parent.go

```diff
@@ -48,13 +49,17 @@ func newParent(env *env) (*parent, map[fileName]*file, error) {
 		}
 	}
 
-	exited := make(chan struct{})
+	exited := make(chan error, 1)
 	go func() {
 		defer rd.Close()
 
-		buf := make([]byte, 1)
-		if _, err := rd.Read(buf); err != io.EOF {
-			panic(err)
+		n, err := io.Copy(ioutil.Discard, rd)
+		if n != 0 {
+			exited <- errors.New("unexpected data tail from parent processdiscarded")
+		}
+		for err != nil {
```

Comment:
I'm not a fan of the infinite loop. How about turning this into an else if, and then add a member to struct Upgrade (protected by upgradeSem) which caches this error?
```suggestion
		else if err != nil {
```